### PR TITLE
[WIP] Add SSL Certificate Support to Fast Pull and Push

### PR DESCRIPTION
**Pull request by ahirreddy**
This PR allows specifying SSL certificates (client key and cert chain) for a specific domain (or set of domains). The transport pool wrapper now exposes a `add_certificate` method that mimics the underlying httplib2.Http transports that it's wrapping.

### CODE REVIEW

**Comment by ahirreddy** [comment]
@mattmoor Does this approach seem reasonable? And should it be expanded to all uses of the transport pool?

Also, I'll open up a pr against docker_rules if/when this guy gets merged. I've got a branch that threads through the arguments.

File: tools/fast_puller_.py [hunk 1]
```diff
@@ -45,6 +45,13 @@
 parser.add_argument('--directory', action='store',
                     help='Where to save the image\'s files.')
 
+parser.add_argument('--certificates', nargs='*', help='A comma separated ' +
```
Comment (ahirreddy) [inline tools/fast_puller_.py]: Maybe this parser argument can get pulled into a common helper since it's duplicated at the moment.

Comment (mattmoor) [reply tools/fast_puller_.py]: Look at what jmillikin did here for logging: https://github.com/google/containerregistry/blob/master/tools/logging_setup_.py#L31

**Comment by mattmoor** [comment]
tl;dr I'm worried this is going to end up making rules very verbose.

What does a `docker_push` look like in the end-state?  What does a `k8s_object` look like (has its own push client)?

An alternative would be to make the client support [something like this](https://docs.docker.com/engine/security/certificates/).  I think my bias is generally towards adopting compatibility with the solutions folks are using with conventional clients, unless there are strong reasons against (maybe there are).

Are you guys using anything like this for certs?

**Comment by KingEmet** [comment]
There's also precedent for defining certs using flags on the Docker client:
https://docs.docker.com/engine/reference/commandline/cli/

**Comment by mattmoor** [comment]
@KingEmet thanks, I was looking for that. :)

**Comment by ahirreddy** [comment]
I'm happy with either solution. Is there a clean way to depend on certs in `/etc/docker/certs.d`? That's actually the solution we use today for Docker pull/push today, we require engineers install those certs.

**Comment by mattmoor** [comment]
@ahirreddy my biggest concern is the `root` requirement for reading that directory :(
```shell
drwx------ 2 root root    4096 Jan 20  2017 docker
```
Since this is daemon-side auth and the daemon runs as root, it can read this ACL'd this way.  We don't (want to) require `root`, so we cannot (as-is).

I think we basically want what you have, but instead of passing the certs it would **enable** reading them from this standard directory (WDYT?).


@KingEmet I don't think there are environment variables controlling this because it's daemon-side.  I think the environment variables control certs for talking to the daemon.  Please correct me if I'm mistaken. :)

**Comment by mattmoor** [comment]
@ahirreddy In summary, if you can also require making that directory user-readable, then the flag option is viable and less verbose.

**Comment by ahirreddy** [comment]
@mattmoor Sounds good! I'll give it a try this weekend

**Comment by KingEmet** [comment]
@mattmoor [looks like it,](https://docs.docker.com/engine/security/https/#create-a-ca-server-and-client-keys-with-openssl) yeah. Seems like `/etc/docker/certs.d/...` is for daemon <-> registry communications

File: tools/fast_puller_.py [hunk 2]
```diff
@@ -58,6 +65,12 @@ def main():
 
   transport = transport_pool.Http(httplib2.Http, size=_THREADS)
 
+  if args.certificates:
+    for item in args.certificates:
+      logging.info('Adding certificate %s', item)
+      key, cert, domain = item.split(',')
+      transport.add_certificate(key, cert, domain)
```
Comment (abergmeier) [inline tools/fast_puller_.py]: Can you please add a test to ensure, this actually works. Looking at the code I have some doubts.

File: tools/fast_pusher_.py [hunk 3]
```diff
@@ -61,6 +61,13 @@
 parser.add_argument('--oci', action='store_true',
                     help='Push the image with an OCI Manifest.')
 
+parser.add_argument('--certificates', nargs='*', help='A comma separated ' +
```
Comment (abergmeier) [inline tools/fast_pusher_.py]: This seems to disallow paths, which contain a comma, right? Any plans for an encoding?

**Comment by abergmeier** [comment]
Um...the merge-base version for this is a bit old *cough*. I am trying to rebase and fix the problems.

**Comment by abergmeier** [comment]
Ok, rebased and pushed to https://github.com/abergmeier/containerregistry/tree/container-push-pull-cert. It now semi-depends on https://github.com/google/containerregistry/pull/51.
@mattmoor @ahirreddy Can you take a look and see whether its good?

**Comment by ahirreddy** [comment]
Sorry, I've been swamped with other things. I'm planning to take a look next week!

**Comment by ahirreddy** [comment]
@mattmoor It occurred to me, it's unclear where to look for the certificates on OS X. Docker for Mac doesn't seem to support client certs, and docker-machine stores them inside the VM.

**Comment by calder** [comment]
@ahirreddy: Any progress on this?
